configserver-client/src/main/java/ai/vespa/hosted/client/ConfigServerClient.java

```diff
@@ -0,0 +1,148 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package ai.vespa.hosted.client;
+
+import org.apache.hc.client5.http.config.RequestConfig;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.core5.http.Method;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toUnmodifiableList;
+
+/**
+ * @author jonmv
+ */
+public interface ConfigServerClient extends AutoCloseable {
+
+    /** Creates a builder for sending the given method, using the specified host strategy. */
+    RequestBuilder send(HostStrategy hosts, Method method);
+
+    /** Builder for a request against a given set of hosts, using this config server client. */
+    interface RequestBuilder {
+
+        /** Sets the request path. */
+        RequestBuilder at(String... pathSegments);
+
+        /** Sets the request body as UTF-8 application/json. */
+        RequestBuilder body(byte[] json);
+
+        /** Sets the request body. */
+        RequestBuilder body(HttpEntity entity);
+
+        /** Sets the parameter key/values for the request. Number of arguments must be even. */
+        RequestBuilder parameters(String... pairs);
+
+        /** Overrides the default socket read timeout of the request. {@code Duration.ZERO} gives infinite timeout. */
+        RequestBuilder timeout(Duration timeout);
+
+        /** Overrides the default socket read timeout of the request. {@code null} allows infinite timeout. */
+        RequestBuilder config(RequestConfig config);
+
+        /**
+         * Sets custom retry/failure logic for this.
+         * <p>
+         * Exactly one of the arguments (response, exception) are non-null.
+         * Return a value to have that returned to the caller;
+         * throw a {@link RetryException} to have the request retried; or
+         * throw any other unchecked exception to have this propagate out to the caller.
+         * The caller must close the provided response, if any.
+         */
+        <T> T handle(BiFunction<ClassicHttpResponse, IOException, T> handler) throws UncheckedIOException;
```

Comment:
How about defining two handlers, one to deal with ClassicHttpResponse and the other to deal with IOException?  Otherwise all handlers would be similar to

```
if (exception == null)
  return handleResponse(response);
else
  return handleException(exception);
```

Reply:
I considered that, but limitations in Java generics then causes you to have to always call one of them before the other, which I think is much more confusing.

Reply:
Uhm, I guess you ocould pass two `Function`s ... yes.

Reply:
Yes, I was thinking of the latter.

Reply:
Perhaps you were thinking of having two of these handle() methods (say `handle` and `handleException`).  That would allow you to define a default behavior with IOException which may be useful?

Reply:
Presumably a RuntimeException could also be caught, and ought to be forwarded to a handle method?

Reply:
The apache client API doesn't list runtime exceptions as part of its API, but states problems and failures are wrapped in an IOException.